python/egobox/egobox.pyi

```diff
@@ -0,0 +1,488 @@
+
+class InfillOptimizer(Enum):
+    """TOTO Infill Optimizer"""
+
+    COBYLA = 1
+    SLSQP = 2
+
+class XType(Enum):
+    """
+    XType is a Rust enum exposed to Python via PyO3.
+
+    Attributes (as uppercase in Python, due to rename_all="UPPERCASE"):
+    - FLOAT
+    - INT
+    - ORD
+    - ENUM
+    """
+
+    FLOAT = 1
+    INT = 2
+    ORD = 3
+    ENUM = 4
+
+class XSpec:
+    """Stub for Rust class XSpec.
+
+    Attributes
+    ----------
+    xtype : XType
+        The type of the x component (e.g., Float)
+    xlimits : list[float]
+        The bounds [lower_bound, upper_bound]
+    tags : list[str]
+        Optional list of tags
+    """
+
+    def __init__(self, xtype: XType, xlimits: list[float], tags: list[str]) -> None: ...
+    @property
+    def xtype(self) -> XType: ...
+    @property
+    def xlimits(self) -> list[float]: ...
+    @property
+    def tags(self) -> list[str]: ...
+
+@dataclass(frozen=True)
+class OptimResult:
+    """
+    A class representing the result of an optimization with associated DOE (Design of Experiments) arrays.
+    This dataclass stub mirrors the read-only attributes exposed by the Rust-backed class.
+    """
+
+    x_opt: NDArray[np.float64]
+    y_opt: NDArray[np.float64]
+    x_doe: NDArray[np.float64]
+    y_doe: NDArray[np.float64]
+
+class ExpectedOptimum:
+    def __init__(self, value: float, tolerance: float = 1e-6) -> None: ...
+    @property
+    def val(self) -> float: ...
+    @property
+    def tol(self) -> float: ...
+
+class Recombination(Enum):
+    """TODO recombination"""
+
+    # prediction is taken from the expert with highest responsability
+    # resulting in a model with discontinuities
+    HARD = 0
+    # Prediction is a combination experts prediction wrt their responsabilities,
+    # an optional heaviside factor might be used control steepness of the change between
+    # experts regions.
+    SMOOTH = 1
+
+def to_specs(xlimits: list[list[float]]) -> list["XSpec"]:
+    """Utility function converting `xlimits` float data list specifying bounds of x components
+    to x specified as a list of XType.Float types `[egobox.XType]`
+
+    Parameters
+    ----------
+    xlimits : list of list of float
+        nx-size list of `[lower_bound, upper_bound]`
+        where `nx` is the dimension of x.
+
+    Returns
+    -------
+    list of XSpec
+        nx-size list of XSpec(XType(Float), [lower_bound, upper_bound])
+        where `nx` is the dimension of x.
+    """
+    ...
+
+class Egor:
+    """Optimizer constructor
+
+    fun: array[n, nx]) -> array[n, ny]
+         the function to be minimized
+         fun(x) = [obj(x), cstr_1(x), ... cstr_k(x)] where
+            obj is the objective function [n, nx] -> [n, 1]
+            cstr_i is the ith constraint function [n, nx] -> [n, 1]
+            an k the number of constraints (n_cstr)
+            hence ny = 1 (obj) + k (cstrs)
+         cstr functions are expected be negative (<=0) at the optimum.
+
+     n_cstr (int):
```

Comment:
This big chunk was also just copied.
I think the stub file is used by the IDE to check the typing and IPython use the runtime inspect check and will get information from pyo3 wrapper. So it has to be duplicated a bit.